core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -951,7 +947,7 @@ private void failOnTragicEvent(AlreadyClosedException ex) {
             failEngine("already closed by tragic event on the index writer", tragedy);
         } else if (translog.isOpen() == false && translog.getTragicException() != null) {
             failEngine("already closed by tragic event on the translog", translog.getTragicException());
-        } else {
+        } else if (failedEngine.get() == null) { // we are closed but the engine is not failed yet?
```

Comment:
Is there a possible thread race condition here, where this thread is handling the `ACE`, but the other thread (that actually closed IW) hasn't been scheduled by the JVM yet and so it didn't get a chance to fail the engine?

Oh I see, this is why you fixed `failEngine` to first set `failedEngine` and THEN close the IW, good!